# Define paths to references
reference = "/lustre1/project/stg_00096/references/GRCh38.alt-masked-V2/index/bwa-mem2/Homo_sapiens_assembly38_masked.fasta"
pod5_dir = "/staging/leuven/stg_00096/home/dkresa/ResolveOME/bulk/1694205563_gc-run-230905-promethion1-fc1d/pod5_pass/"
dorado_dir = "/staging/leuven/stg_00096/home/dkresa/singularity/dorado_v0.4.1.sif"
stage_dir = "/staging/leuven/stg_00096"
lustre_dir = "/lustre1/project/stg_00096"
annotation = "/staging/leuven/stg_00096/home/dkresa/data/reference/annotations/gencode.v44.basic.annotation.bed"

# Define output files
dorado_basecaller_output = "basecalled/bam/1694205563_basecalled.bam"
minimap2_output = "basecalled/bam/1694205563_basecalled_aligned.primary.bam"
sorted_output = "bams/1694205563.sorted.primary.bam.bai"

#Define output reports
dorado_basecaller_report = "reports/dorado_basecaller/1694205563_basecalled.txt"
minimap2_report = "reports/dorado_basecaller/1694205563_basecalled_aligned.primary.txt"
sorted_report = "reports/sorted/1694205563.sorted.primary.txt"


rule all:
    input:
        dorado_basecaller_output,
        dorado_basecaller_report,
        minimap2_output,
        minimap2_report,
        sorted_output,
        sorted_report


# Rule for basecalling in dorado
rule dorado_basecaller:
    input:
        pod5 = pod5_dir,
        ref = reference
    output:
        bam = "basecalled/bam/1694205563_basecalled.bam",
        task_done = "reports/dorado_basecaller/1694205563_basecalled.txt"
    params:
        stage = stage_dir,
        lustre = lustre_dir,
        dorado = dorado_dir
    threads: 24
    shell:
        """
        singularity run --nv \
        -B {params.stage} \
        -B {params.lustre} \
        {params.dorado} \
        /opt/dorado/bin/dorado basecaller \
        /opt/dorado/bin/dna_r10.4.1_e8.2_400bps_sup@v4.2.0 {input.pod5} \
        --kit-name SQK-NBD114-24 \
        > {output.bam}

        touch {output.task_done}
        """

# Rule for minimap2 alignment to the reference genome
rule minimap2:
    input:
        report_files = dorado_basecaller_report,
        unmapped = "basecalled/bam/1694205563_basecalled.bam",
        ref = reference,
        anno = annotation
    threads: 24
    output:
        bam = "basecalled/bam/1694205563_basecalled_aligned.primary.bam",
        task_done = "reports/dorado_basecaller/1694205563_basecalled_aligned.primary.txt"
    shell:
        """
        minimap2 -ax splice -uf -k14 --secondary=no --junc-bed {input.anno} -t {threads} {input.ref} {input.fq} | samtools view -Sb > {output.bam}
        
touch {output.task_done}
        """

# Rule for sorting and indexing BAM files
rule sort_and_index:
    input:
        report_files = minimap2_report,
        bam = "basecalled/bam/1694205563_basecalled_aligned.primary.bam"
    output:
        sorted_bam = "bams/1694205563.sorted.primary.bam",
        bam_index = "bams/1694205563.sorted.primary.bam.bai",
        task_done = "reports/sorted/1694205563.sorted.primary.txt"
    threads: 24
    shell:
        """
        samtools sort -@ {threads} -o {output.sorted_bam} {input.bam}
        samtools index {output.sorted_bam} {output.bam_index}

        touch {output.task_done}
        """
